rust: monitor factory: add `monitor facotry` in RROS

shannmu opened the pull request:
The ` monitor factory` has implemented a synchronization mechanism inter threads, providing three types of synchronization methods to the user mode: mutexes, semaphores, and condition variables. During the implementation process, I believe it is worth mentioning that the following features were added or modified:

- Fixed bugs in `mutex.rs`.

- Added a `u_window` field to the RrosThread structure, which is used for kernel mode and user mode to interact data through shared memory.

- As mentioned above, the interaction of some data between kernel mode and user mode adopts shared memory. This change fixed a bug in control.rs, which uses `mmap` to create shared memory; and it implemented a memory allocator for shared memory.

- Added the setting of thread FIFO scheduling strategy and fixed some functions related to thread scheduling.

https://github.com/BUPT-OS/RROS/issues/24

shannmu committed:
rust: monitor factory: add `monitor facotry` in RROS

The ` monitor factory` has implemented a synchronization mechanism inter threads, providing three types of synchronization methods to the user mode: mutexes, semaphores, and condition variables. During the implementation process, I believe it is worth mentioning that the following features were added or modified:

- Fixed bugs in `mutex.rs`.

- Added a `u_window` field to the RrosThread structure, which is used for kernel mode and user mode to interact data through shared memory.

- As mentioned above, the interaction of some data between kernel mode and user mode adopts shared memory. This change fixed a bug in control.rs, which uses `mmap` to create shared memory; and it implemented a memory allocator for shared memory.

- Added the setting of thread FIFO scheduling strategy and fixed some functions related to thread scheduling.

https://github.com/BUPT-OS/RROS/issues/24

Lukeehoojet commented:
The CI passed, but there are many warnings in your code, try to eliminate them before submitting.

shannmu committed:
rust: monitor: fix the warnings.

This commit is a warnings clean commit.

Lukeehoojet commented on kernel/rros/thread.rs:
```diff
@@ -1739,8 +1884,9 @@ fn thread_factory_build(
     // 	tsk->comm[sizeof(tsk->comm) - 1] = '\0';
-    // `RrosElement.pointer` point to `rros_thread` struct.
-    let rros_thread_ptr: *mut RrosThread = curr.locked_data().get();
+    // `RrosElement.pointer` point to `RrosThread` struct.
+    let rros_thread_ptr =
+        Arc::as_ptr(&curr) as *const SpinLock<RrosThread> as *mut SpinLock<RrosThread>;
     (curr.lock().element.borrow_mut()).pointer = rros_thread_ptr as *mut u8;
```
Does it have to be `*mut SpinLock<RrosThread>` rather than `*mut RrosThread`?

shannmu replied:
I think this code must be implemented like this. `element.pointer` is the only value that could get back the element itself. If we use `*mut RrosThraed`, it will casue a memory leak on the `Spinlock`, and we can't get back the `Spinlock`.

shannmu committed:
rust: monitor: add dispose method to monitor factory

This method is called when the monitor is no longer needed. It should be used to clean up any resources that the monitor may have allocated.

shannmu committed:
rust: rros: fix the error and warning producing by rebase operation

- Use `rros_get_element_by_fundle` to replace `__rros_get_element_by_fundle`

- Use new `lock::hard_xx` to replace `lock::raw_xx` in mutex.rs

ruiqurm reviewed (approved):
LGFM
